src/flb_engine.c

```diff
@@ -514,6 +517,7 @@ int flb_engine_start(struct flb_config *config)
     /* Start the Storage engine */
     ret = flb_storage_create(config);
     if (ret == -1) {
+        flb_error("[engine] storage creation failed");
```

Comment:
after this line all the flb_ logging family function are available:

https://github.com/fluent/fluent-bit/pull/3723/files#diff-8bae5597d65353bb9faee8e5c4d636ba1e06740937fb0cf3e6dc8cfcc43c3b43R494

if something fails and the logger don't print anything, can we take a look at the root cause of that ?

Reply:
I looked through flb_log_create and it does report any errors it encounters either via flb_errno or fprintf, which is why I use fprintf for the first few failures in flb_engine_start.

Unfortunately flb_errno calls flb_errno_print which calls flb_error...

```c
int flb_errno_print(int errnum, const char *file, int line)
{
    char buf[256];

    strerror_r(errnum, buf, sizeof(buf) - 1);
    flb_error("[%s:%i errno=%i] %s", file, line, errnum, buf);
    return 0;
}
```

That may be the cause for silent failures when there are logging errors. Should we store a thread local global boolean flag to signal when the log system has been initialized to try and solve that?

Reply:
looking through further, flb_log_print actually falls back on printing to stderr when it cannot acquire the fluent-bit worker context via flb_worker_get. If it fails to write to the log pipe via flb_pipe_write_all it will report the write error via perror but not report the actual error log.

Reply:
I added a call to fprintf before the line that resorts to perror, which I realize now is redundant. 

There's a part of me that is convinced we should probably be using read(2) with STDOUT_FILENO as our fall back instead of stdio.

Reply:
This PR is pretty much done for now.